packages/mock-block-dock/src/datastore/traverse/traversal-context.ts

```diff
@@ -0,0 +1,96 @@
+import {
+  EntityEditionId,
+  GraphResolveDepths,
+  OntologyTypeEditionId,
+  Subgraph,
+} from "@blockprotocol/graph";
+
+import { typedEntries } from "../../util";
+import { ResolveMap } from "./resolve-map";
+
+export type PartialDepths = {
+  [K1 in keyof GraphResolveDepths]?: {
+    [K2 in keyof GraphResolveDepths[K1]]?: GraphResolveDepths[K1][K2];
+  };
+};
+
+export class Depths {
+  constructor(public inner: PartialDepths) {}
+
+  /**
+   * Takes a subset of `GraphResolveDepths` and checks if any of the supplied depths are greater than the ones already
+   * tracked in this. If they are, this object is updated and the different depths are returned.
+   *
+   * @param {PartialDepths} newDepths - the new subset of depths at which the element should be resolved to
+   * @returns {PartialDepths} - the depths that were greater in `newDepths` than they were in `this`
+   */
+  update(newDepths: PartialDepths): PartialDepths {
+    const updatedDepths: PartialDepths = {};
+    for (const [edgeKind, directions] of typedEntries(newDepths)) {
+      if (directions) {
+        for (const [direction, depth] of typedEntries(directions)) {
+          if (depth < 1) {
+            continue;
+          }
+          if (!this.inner[edgeKind]) {
+            this.inner[edgeKind] = {
+              [direction]: depth,
+            };
+          } else if (!this.inner[edgeKind]![direction]) {
+            this.inner[edgeKind]![direction] = depth;
+          } else if (depth > this.inner[edgeKind]![direction]!) {
+            this.inner[edgeKind]![direction] = depth;
+          } else {
+            continue;
+          }
+
+          updatedDepths[edgeKind] = {
+            ...updatedDepths[edgeKind],
+            [direction]: depth,
+          };
+        }
+      }
+    }
+
+    return updatedDepths;
+  }
+}
+
+/**
+ * Tracks the depths at which each element within the graph has been explored. This means that if the same element is
+ * encountered multiple times during traversal, at different respective depths, the context can determine if its
+ * neighbours have already been explored to a deeper depth than the current branch of the traversal, or if further
+ * exploration is needed.
+ */
+export class TraversalContext {
```

Comment:
This is very similar to our `DependencyContext` in the HASH graph implementation. Should probably try and take some lessons from the way I did it here and backport it back. Already discussing with @TimDiekmann about some of them